### PR TITLE
feat: support ESLint v8

### LINKED ISSUE #728: Support ESLint 8
⭐ Support ESLint 8.

This is a **tracking issue** for supporting `eslint@8`. It's currently being blocked by:

- [x] `typescript-eslint` ([now supports `eslint@8`](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v5.0.0))
- [x] `eslint-config-standard` (https://github.com/standard/eslint-config-standard/pull/193)

---

New `eslint` rules with `typescript-eslint` equivalents:
- [no-loss-of-precision](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-loss-of-precision.md)

New rules worth considering:
- [no-unnecessary-type-constraint](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md)
- [no-unsafe-argument](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unsafe-argument.md)

---

Refs: 
- [Changes to the recommended sets for 5.0.0](https://github.com/typescript-eslint/typescript-eslint/issues/3746)

### PR DESCRIPTION
**Pull request by Divlo**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix
[x] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Support ESLint v8.

**Which issue (if any) does this pull request address?**

Fixes #728

**Is there anything you'd like reviewers to focus on?**

Even if #727, @mightyiam, do you still want to review or we can merge/release with another standard maintainer approval?

### CODE REVIEW

**Comment by jefbarn** [comment]
Possible to publish a pre-release version of this for testing? (because the config file is typescript, can't just use the branch directly in eslint config)

**Comment by Divlo** [comment]
@jefbarn Yes, we can certainly do a prerelease before a stable release.
I don't have the right to publish on npm, we need to wait an approval from another `standard` maintainer.

File: package.json
Comment (mightyiam) [inline package.json:66]: Is this the earliest release we are able to require?

File: package.json
Comment (mightyiam) [inline package.json:67]: Is this the earliest release we are able to require?

File: readme.md
Comment (mightyiam) [inline readme.md:52]: Just `Example command line usage:` perhaps?

**Review by mightyiam — CHANGES_REQUESTED** [review]

File: package.json
Comment (Divlo) [reply package.json:8]: It is what ESLint >= 8.0.0 support:
Ref: https://github.com/eslint/eslint/blob/6f940c3ce715327f282c197d0f71b91848e5d83d/package.json#L146-L148

File: package.json
Comment (Divlo) [reply package.json:60]: It is a prerelease, `17.0.0` doesn't exist yet:
Ref: https://github.com/standard/eslint-config-standard/issues/208

File: package.json
Comment (Divlo) [reply package.json:65]: Yes, see: https://github.com/standard/eslint-config-standard/blob/63e644c7a2bf464f56afad1324bd88ea9cc6f6a6/package.json#L52-L55

File: package.json
Comment (Divlo) [reply package.json:66]: Yes, see: https://github.com/standard/eslint-config-standard/blob/63e644c7a2bf464f56afad1324bd88ea9cc6f6a6/package.json#L52-L55

File: package.json
Comment (Divlo) [reply package.json:67]: Yes, see: https://github.com/standard/eslint-config-standard/blob/63e644c7a2bf464f56afad1324bd88ea9cc6f6a6/package.json#L52-L55

File: package.json
Comment (mightyiam) [reply package.json:8]: Still seems like an unrelated change, doesn't it? Wouldn't this PR be complete without it?

File: package.json
Comment (mightyiam) [reply package.json:60]: This opens a discussion about prereleases. I'm not sure I'd like to go there.

File: package.json
Comment (mightyiam) [reply package.json:64]: Right.

File: package.json
Comment (mightyiam) [reply package.json:65]: I'm happy about these tests, yes.

File: package.json
Comment (mightyiam) [reply package.json:67]: That test makes these questions unnecessary.

File: package.json
Comment (Divlo) [reply package.json:8]: Yes, we can merge the PR without this, but at least it would be clearer for users what Node.js version we support.
I can revert this change if you really want to.

File: package.json
Comment (Divlo) [reply package.json:60]: We don't need to do prereleases if you don't want to, and that's totally fine.
As said in the PR description:
> Should not be merged/released until `eslint-config-standard@17` has been released stable.

So the only thing that we'll need to change in this PR once released is this line to be: `"eslint-config-standard": "^17.0.0"`.

File: package.json
Comment (Divlo) [reply package.json:60]: I don't remember exactly why we needed to change this to be `^4.0.1 || ^5.0.0`, but to support ESLint >= 8.0.0, we must use `"@typescript-eslint/eslint-plugin": "^5.0.0"` .

Related changelog: https://github.com/typescript-eslint/typescript-eslint/blob/main/CHANGELOG.md#500-2021-10-11

File: package.json
Comment (mightyiam) [reply package.json:8]: Yes, please.

File: package.json
Comment (mightyiam) [reply package.json:60]: Yes, defer merging until upstream release.

File: package.json
Comment (mightyiam) [reply package.json:60]: Perfect.

File: package.json
Comment (jefbarn) [reply package.json:60]: @mightyiam There have been several requests to do a pre-release in the ESLint v8 issue. It appears the upstream will be in pre-release for a while. Can we please also get a pre-release of this package?

Comment (voxpelli) [reply package.json:60]: This becomes a bit of a catch-22, as the `standard` prerelease is to allow projects to give feedback on regressions and/or missing features (like https://github.com/standard/standard/issues/1548)

Without a prerelease of this module, that excludes the users of `eslint-config-standard-with-typescript` from contributing to that.

Not sure why a prerelease would be a bad thing?

**Review by voxpelli — COMMENTED** [review]
I'm going to leave it to @mightyiam to decide on this one, as he is the one driving this part of `standard`

File: package.json
Comment (mightyiam) [reply package.json:60]: Pre-release adds some complexity. A pre-release should be developed on a branch of its own, for one.

I am not aware of the value that users find in pre-releases. Would anyone care to explain?

Comment (voxpelli) [reply package.json:60]: > A pre-release should be developed on a branch of its own, for one.

I would say that the main branch is where all development happen, no matter if it’s ending up in a stable release or a prerelease.

Separate branches as I see it is only used for maintaining legacy versions?

> I am not aware of the value that users find in pre-releases. Would anyone care to explain?

When we release a new major version we commit to a set of breaking changes. If those breaking changes doesn’t align with our intended breaking changes or the changes desired by the community, then we have to roll yet another major release to fix that.

Prereleases enables us + interested parties in the community to test out a new major before we commit to the breaking changes, making it possible for us to ensure that we ship what we intend to ship.

It also allows those in the community who feels blocked by current major version to unblock themselves earlier than would otherwise be possible.

Comment (mightyiam) [reply package.json:60]: > > A pre-release should be developed on a branch of its own, for one.
> 
> I would say that the main branch is where all development happen, no matter if it’s ending up in a stable release or a prerelease.
> 
> Separate branches as I see it is only used for maintaining legacy versions?

Would you like the main branch to always be releasable?
> 
> > I am not aware of the value that users find in pre-releases. Would anyone care to explain?
> 
> When we release a new major version we commit to a set of breaking changes. If those breaking changes doesn’t align with our intended breaking changes or the changes desired by the community, then we have to roll yet another major release to fix that.
> 
> Prereleases enables us + interested parties in the community to test out a new major before we commit to the breaking changes, making it possible for us to ensure that we ship what we intend to ship.
> 
> It also allows those in the community who feels blocked by current major version to unblock themselves earlier than would otherwise be possible.

Thank you. This last paragraph I didn't understand. What is "current major version" in this scenario?

**Comment by flippidippi** [comment]
Any updates on this? We are waiting for this to be released so we can upgrade to eslint 8.

**Comment by Divlo** [comment]
> Any updates on this? We are waiting for this to be released so we can upgrade to eslint 8.

@flippidippi It is blocked by: https://github.com/standard/eslint-config-standard/issues/208

**Comment by jefbarn** [comment]
For people wanting to start using this, my solution was:
1. Fork the repo
2. Switch to this branch ```git branch feat/eslint-support-v8```
3. Compile the code ```npm run compile```
4. Remove '/lib' from .gitignore
5. Commit and push the /lib folder to your new repo
6. Use the project.json github syntax:
```"eslint-config-standard-with-typescript": "jefbarn/eslint-config-standard-with-typescript#feat/support-eslint-v8"```

Though this solution is not great if we need to merge in changes from upstream.

File: package.json
Comment (Divlo) [reply package.json]: Updated. :+1:

File: package.json
Comment (Divlo) [reply package.json]: `eslint-config-standard@17` has been released stable now. :tada:

**Comment by Divlo** [comment]
@mightyiam This should now be ready! :smile:
Do you mind reviewing this PR again and if everything is fine, releasing a new major version of `eslint-config-standard-with-typescript`, thanks!

**Comment by mightyiam** [comment]
I'll do my best.

**Comment by gutenye** [comment]
Any updates on the release? [eslint-config-standard](https://github.com/standard/eslint-config-standard/issues/208#issuecomment-1104059454) is released

**Pull request merged**
